### ISSUE TITLE
Expired key - "Google Cloud Packages Automatic Signing Key"

### ISSUE THREAD

**Issue by david-caro** [kubernetes/release#1982]
The current available key expired yesterday:
```
dcaro@vulcanus$ gpg --list-key 6A030B21BA07F4FB
pub   rsa2048 2018-04-01 [SCE] [expired: 2021-03-31]
      54A647F9048D5688D7DA2ABE6A030B21BA07F4FB
uid           [ expired] Google Cloud Packages Automatic Signing Key <gc-team@google.com>
```

**What happened**: The key used to sign all the packages from the packages.cloud.google.com repos `https://packages.cloud.google.com/apt/dists/debian-buster-mirror/Release.gpg` has expired.

**What you expected to happen**: The key to be refreshed :)

**How to reproduce it (as minimally and precisely as possible)**: 

**Please provide links to example occurrences, if any**: This is a new occurrence of  kubernetes/test-infra#7505

**Anything else we need to know?**: Thanks for all the good work!

**Comment by spiffxp** [comment]
/kind bug
/sig release
/priority critical-urgent
re-applying labels

**Comment by spiffxp** [comment]
/milestone v1.21
since I assume this may impact cutting 1.21 packages

**Comment by BenTheElder** [comment]
I no longer seen an expired key when importing it in debian, there is now this published as well: https://cloud.google.com/compute/docs/troubleshooting/known-issues#keyexpired

**Comment by BenTheElder** [comment]
https://cloud.google.com/compute/docs/troubleshooting/known-issues#keyexpired should probably be shared somewhere (perhaps the kubernetes.io install docs?) but I don't think there's anything else further to do here.

The keys have been updated, yum continues to have problematic behavior here but the workaround is published in the link above.

**Comment by spiffxp** [comment]
/remove-priority critical-urgent
/priority important-longterm

Downgrading priority. If we don't hear back from the author of this issue I would recommend we close unless anyone has any concrete follow ups they want to hold open for.

**Comment by rhino5oh** [comment]
This is still broken for people on at least CentOS 7. I dont think the priority should be downgraded. I dont know what the exact solution would be, but I dont think disabling a gpg check is the permanent solution.

**Comment by spiffxp** [comment]
Have you confirmed the steps at https://cloud.google.com/compute/docs/troubleshooting/known-issues#keyexpired repro the issue for you? Does your repo definitely use gpg instead of https for verification?

(I haven't checked on internal issue to confirm where rollout of the key is for yum, this info would help)

**Comment by david-caro** [comment]
This worked for me, though we are using the key with reprepro to mirror the repo, so not using the repo directly.

**Comment by BenTheElder** [comment]
> I dont think the priority should be downgraded.

To be clear the Kubernetes project does not own this apt/yum hosting nor does the Google team liasing. We can't do anything further about it except passing along the concern(s).

Further, there is a workable workaround now, and as noted in the guide having the signature here is unusual and poorly supported in the yum tooling (it assumes a single key ever which is obviously problematic ...).

The issue referenced above is an existing issue about moving to community managed hosting, which could use contributors. https://github.com/kubernetes/kubernetes/issues/100348
If this was solved we could change the repos to our whims. As it stands we piggyback on infrastructure that is not publicly managed and not offered as a product.

**Comment by RakhithaRR** [comment]
> Same as [kubernetes/test-infra#7505 (comment)](https://github.com/kubernetes/test-infra/issues/7505#issuecomment-378045292)
> 
> There was a new key pushed, it may be all you need to do is update to use the new signing key
> 
> ```
> $ curl -s https://packages.cloud.google.com/apt/doc/apt-key.gpg | gpg -v -
> gpg: WARNING: no command supplied.  Trying to guess what you mean ...
> pub   rsa2048 2018-04-01 [SCE] [expired: 2021-03-31]
>       54A647F9048D5688D7DA2ABE6A030B21BA07F4FB
> uid           Google Cloud Packages Automatic Signing Key <gc-team@google.com>
> sig        6A030B21BA07F4FB 2018-04-01   [selfsig]
> pub   rsa2048 2020-12-04 [SC] [expires: 2022-12-04]
>       59FE0256827269DC81578F928B57C5C2836F4BEB
> uid           gLinux Rapture Automatic Signing Key (//depot/google3/production/borg/cloud-rapture/keys/cloud-rapture-pubkeys/cloud-rapture-signing-key-2020-12-03-16_08_05.pub) <glinux-team@google.com>
> sig        8B57C5C2836F4BEB 2020-12-04   [selfsig]
> sub   rsa2048 2020-12-04 [E]
> sig        8B57C5C2836F4BEB 2020-12-04   [keybind]
> ```
> 
> Something like this should work?
> 
> ```
> curl -s https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -
> ```
> 
> or
> 
> ```
> sudo apt-key adv --keyserver keyserver.ubuntu.com --recv 8B57C5C2836F4BEB
> ```
> 
> Having said that, the appropriate team internally is currently fielding a number of issues. I believe the new key may not have fully rolled out yet. If the above doesn't work for you, please report back.

I see a similar issue in Ubuntu 20.04.

`apt-get update` shows this error 
```
An error occurred during the signature verification. The repository is not updated and the previous index files will be used. GPG error: https://packages.cloud.google.com/apt cloud-sdk InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY FEEA9169307EA071 NO_PUBKEY 8B57C5C2836F4BEB
```
Running the above two commands you've mentioned did not solve this.
I can still see that `Google Cloud Packages Automatic Signing Key` is expired (on 2021-03-31) when I run `apt-key list`.

**Comment by laurenegerton** [comment]
I see the same expiration as @RakhithaRR.

**Comment by Greg0** [comment]
Same issue on my machine while `apt update`

```
W: An error occurred during the signature verification. The repository is not updated and the previous index files will be used. GPG error: https://packages.cloud.google.com/apt kubernetes-xenial InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY FEEA9169307EA071 NO_PUBKEY 8B57C5C2836F4BEB
```

**Comment by bharrisaurora** [comment]
> Same issue on my machine while `apt update`
> 
> ```
> W: An error occurred during the signature verification. The repository is not updated and the previous index files will be used. GPG error: https://packages.cloud.google.com/apt kubernetes-xenial InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY FEEA9169307EA071 NO_PUBKEY 8B57C5C2836F4BEB
> ```

I am also seeing this.  This seems to happen pretty regularly to be honest.  I'm not sure why it's such a struggle to keep the needed keys available to download.

**Comment by dims** [comment]
@kubernetes/release-team-leads can you please see if there are ways for us to avoid/reduce this barrier in the future?

**Comment by justaugustus** [comment]
> @kubernetes/release-team-leads can you please see if there are ways for us to avoid/reduce this barrier in the future?

cc: @kubernetes/release-engineering @kubernetes/build-admins @kubernetes/sig-release-leads 

(Since this is RelEng)

**Comment by saschagrunert** [comment]
> I see a similar issue in Ubuntu 20.04.
> 
> `apt-get update` shows this error
> 
> ```
> An error occurred during the signature verification. The repository is not updated and the previous index files will be used. GPG error: https://packages.cloud.google.com/apt cloud-sdk InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY FEEA9169307EA071 NO_PUBKEY 8B57C5C2836F4BEB
> ```
> 
> Running the above two commands you've mentioned did not solve this.
> I can still see that `Google Cloud Packages Automatic Signing Key` is expired (on 2021-03-31) when I run `apt-key list`.

Can you confirm that something like this works for you or provide a more encapsulated reproducer:

```Dockerfile
FROM ubuntu:20.04

RUN apt-get update
RUN apt-get install -y curl gpg

RUN curl -fsSLo /usr/share/keyrings/kubernetes-keyring.gpg https://packages.cloud.google.com/apt/doc/apt-key.gpg
RUN echo "deb [signed-by=/usr/share/keyrings/kubernetes-keyring.gpg] https://apt.kubernetes.io/ kubernetes-xenial main" | tee /etc/apt/sources.list.d/kubernetes.list

RUN apt-get update
RUN apt-get install -y kubelet
```

**Comment by jsoref** [comment]
The problem with just adding the key is that you'll still get this: https://github.com/kubernetes/release/issues/1982#issuecomment-826582877

because the odds are tat your repository has pinned the key:
```
/etc/apt/sources.list.d/google-cloud-sdk.list:deb [signed-by=/usr/share/keyrings/cloud.google.gpg] https://packages.cloud.google.com/apt cloud-sdk main
```

1. Someone (possibly me) needs to file a bug (and possibly a patch) to apt to change the error message:

    ```diff
    -The following signatures couldn't be verified because the public key is not available
    +The following signatures couldn't be verified because the public key does not match the keys listed in {file}
    ```

2. Everyone else just needs to either edit the `/etc/apt/sources.list.d/google-cloud-sdk.list` file to change the `signed-by` field (or remove it), or replace the contents of that file w/ the updated value

3. If at all possible, someone from google should provide a way to upgrade that file.

**Comment by contributorpw** [comment]
OK. This was killing me for a while. I have to do this on Ubuntu 20.04

```sh
$> curl https://packages.cloud.google.com/apt/doc/apt-key.gpg -o /usr/share/keyrings/cloud.google.gpg
```

Try before this
```sh
$> apt-key del FEEA9169307EA071 && apt-key del 8B57C5C2836F4BEB
```

**Comment by aleon1220** [comment]
I will share a command that worked for me in **Ubuntu 18.04** 

``` bash
curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | sudo apt-key add -
```

Source [Known Issues Google Cloud](https://cloud.google.com/compute/docs/troubleshooting/known-issues#keyexpired)

**Comment by jsoref** [comment]
Note that `apt-key` is deprecated -- https://www.linuxuprising.com/2021/01/apt-key-is-deprecated-how-to-add.html --  so you really shouldn't recommend using it (this is more or less why I don't).

**Comment by Slyke** [comment]
@jsoref what should we use instead of this command?
```
curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | sudo apt-key add -
```

**Comment by contributorpw** [comment]
> @jsoref what should we use instead of this command?
> 
> ```
> curl ... | ... apt-key add -
> ```

Just put the key to the end directory https://github.com/kubernetes/release/issues/1982#issuecomment-883865752

**Comment by jsoref** [comment]
You put it in the appropriate directory (usually trusted.gpg.d) with the appropriate file extension (.gpg for binary and .asc for the older armored text format). 

https://github.com/kubernetes/release/issues/1982#issuecomment-838930333

**Comment by Slyke** [comment]
So:
```
curl https://packages.cloud.google.com/apt/doc/apt-key.gpg --output /etc/apt/trusted.gpg/apt-key.gpg
```?

**Comment by jsoref** [comment]
It depends on which edge case you're running into. The file extension absolutely matters. If it's a binary (and that one is), then `.gpg` is required, if it's text, then `.asc` is required. If you have a repository which has pinned its key to a specific absolute path on your file system, then you need to dump the key exactly there, otherwise, no the filename doesn't matter.

**Comment by Slyke** [comment]
Yep, this seems to be working for me:
```
sudo curl https://packages.cloud.google.com/apt/doc/apt-key.gpg --output /etc/apt/trusted.gpg.d/k8s-apt-key.gpg # This line was apt-key
sudo apt-add-repository "deb http://apt.kubernetes.io/ kubernetes-xenial main" -y
sudo apt update
sudo apt install kubeadm kubelet kubectl -y || { echo 'Failed to install kubeadm kubelet kubectl. Please retry' ; exit 1; }
```

Note that I had to use `/etc/apt/trusted.gpg.d/k8s-apt-key.gpg` and not `/etc/apt/trusted.gpg/k8s-apt-key.gpg`. The folder `/etc/apt/trusted.gpg` didn't exist on my system. This is part of a larger bash script that installs K8s on fresh systems.

**Comment by jsoref** [comment]
Oh, sorry, I missed that. Yes, the debian* design uses `.d` to indicate a directory that can contain multiple files that are collectively used as if their contents were part of the single file without the `.d`.

* Probably used elsewhere too…

**Comment by JamesXNelson** [comment]
@contributorpw already answered it and shared perfect workaround for me here: curl https://packages.cloud.google.com/apt/doc/apt-key.gpg -o /usr/share/keyrings/cloud.google.gpg

(the answer above with hearts on it, go add one if it works for you!)

**Comment by korcky** [comment]
So, I tried to follow a simple instruction on how to [install kubernetes using native package management](https://kubernetes.io/docs/tasks/tools/install-kubectl-linux/#install-using-native-package-management) for Debian-based distributions (my os version is `Ubuntu 20.04`) and get some of the errors mentioned above. To be precise, `apt update` output this lines:
```
Err:8 https://packages.cloud.google.com/apt kubernetes-xenial InRelease                          
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY FEEA9169307EA071 NO_PUBKEY 8B57C5C2836F4BEB
...
Reading package lists... Done
W: GPG error: https://packages.cloud.google.com/apt kubernetes-xenial InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY FEEA9169307EA071 NO_PUBKEY 8B57C5C2836F4BEB
E: The repository 'https://apt.kubernetes.io kubernetes-xenial InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
```
As were mentioned [#1982 (comment)](https://github.com/kubernetes/release/issues/1982#issuecomment-838902947), I tried to remove `signed-by` part (but in `/etc/apt/sources.list.d/kubernetes.list`) and added one of the missing key with `apt-key` and it worked.
To summarize, the solution in my case was to change a little be the instruction mentioned above:
1. `sudo apt-get update && sudo apt-get install -y apt-transport-https ca-certificates curl`
2. `echo "deb https://apt.kubernetes.io/ kubernetes-xenial main" | sudo tee /etc/apt/sources.list.d/kubernetes.list`
3. `sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 8B57C5C2836F4BEB`
4. `sudo apt-get update && sudo apt-get install -y kubectl`

**Comment by rmb938** [comment]
Did the keys expire again? Getting issues trying to isntall things on CentOS

```
# yum -q makecache -y --disablerepo=\* --enablerepo=kubernetes
Importing GPG key 0x7AA99B80:
 Userid     : "Rapture Automatic Signing Key (cloud-rapture-signing-key-2022-03-07-08_01_01.pub)"
 Fingerprint: e89b faec 34d3 1829 362f 7f7b 5668 44f3 7aa9 9b80
 From       : https://packages.cloud.google.com/yum/doc/yum-key.gpg
https://packages.cloud.google.com/yum/repos/kubernetes-el7-x86_64/repodata/repomd.xml: [Errno -1] repomd.xml signature could not be verified for kubernetes
```

**Comment by aleksandar-kinanov** [comment]
Hitting the same issue on CentOS 7:
```
failure: repodata/repomd.xml from kubernetes: [Errno 256] No more mirrors to try.
https://packages.cloud.google.com/yum/repos/kubernetes-el7-x86_64/repodata/repomd.xml: [Errno -1] repomd.xml signature could not be verified for kubernetes
```

**Comment by BenTheElder** [comment]
Folks, the Kubernetes project does not manage these keys, nor does the team that voluntarily continues to publish these packages until the project moves to community infrastructure. https://github.com/kubernetes/release/issues/913 seems to be the current tracking issue for this.

per https://cloud.google.com/compute/docs/troubleshooting/known-issues#keyexpired the recommendation is to disable GPG check in RPM.

> To fix this, disable repository GPG key checking in the yum repo configuration by setting repo_gpgcheck=0. In supported Compute Engine base images this setting might be found in /etc/yum.repos.d/google-cloud.repo file. However, your VM can have this set in different repository configuration files or automation tools.
>
> Yum repositories do not usually use GPG keys for repository validation. Instead, the https endpoint is trusted.

**Comment by tallaxes** [comment]
Another reason for getting NO_PUBKEY on `apt-get update` is when the GPG key file is not world-readable - easy to end up like this by accident with tight `umask` defaults. And if you are wondering (as I was) why it needs to be world-readable even when apt runs as root - it is because apt switches to `_apt` user for many operations that don't require privilege (see [here](https://askubuntu.com/questions/1401907/why-must-apt-gpg-keys-be-world-readable)). 

A simple `chmod o+r <keyfile>` fixes NO_PUBKEY in this particular case.

**Comment by contributorpw** [comment]
> A simple `chmod o+r <keyfile>` fixes NO_PUBKEY in this particular case.

@tallaxes 

It might make sense. This is probably the main reason. Thanks!

**Comment by sftim** [comment]
Docs updated in https://github.com/kubernetes/website/pull/34546 (relevant to Yum; not relevant to APT)

**Comment by sftim** [comment]
https://cloud.google.com/compute/docs/troubleshooting/known-issues#keyexpired-2 is still relevant advice. https://github.com/kubernetes/release/issues/913 tracks migrating to Kubernetes' own packaging and repos.

I'm minded to close this now.

**Comment by myself659** [comment]
https://cloud.google.com/compute/docs/troubleshooting/known-issues#ubuntu-systems 

```
curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | sudo apt-key --keyring /usr/share/keyrings/cloud.google.gpg add -
```
this works for me

**Comment by GannonTdW** [comment]
Debian 11 bullseye is supported or not ? 
I have a issue with the gpg key:

```
root@k8s-worker-111:~# cat /etc/*release*
ID="genericcloud"
VERSION="20220711-1073"
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

root@k8s-worker-111:~# curl -fsSLo /usr/share/keyrings/kubernetes-archive-keyring.gpg https://packages.cloud.google.com/apt/doc/apt-key.gpg

root@k8s-worker-111:~# echo "deb [signed-by=/usr/share/keyrings/kubernetes-archive-keyring.gpg] https://apt.kubernetes.io/ kubernetes-xenial main" | tee /etc/apt/sources.list.d/kubernetes.list
deb [signed-by=/usr/share/keyrings/kubernetes-archive-keyring.gpg] https://apt.kubernetes.io/ kubernetes-xenial main

root@k8s-worker-111:~# apt-get update
Hit:1 http://security.debian.org/debian-security bullseye-security InRelease
Hit:2 https://download.docker.com/linux/debian bullseye InRelease                  
Hit:3 http://deb.debian.org/debian bullseye InRelease                              
Hit:4 http://deb.debian.org/debian bullseye-updates InRelease
Hit:5 http://deb.debian.org/debian bullseye-backports InRelease
Get:6 https://packages.cloud.google.com/apt kubernetes-xenial InRelease [9383 B]
Err:6 https://packages.cloud.google.com/apt kubernetes-xenial InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY FEEA9169307EA071 NO_PUBKEY 8B57C5C2836F4BEB
Reading package lists... Done
W: GPG error: https://packages.cloud.google.com/apt kubernetes-xenial InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY FEEA9169307EA071 NO_PUBKEY 8B57C5C2836F4BEB
E: The repository 'https://apt.kubernetes.io kubernetes-xenial InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.

root@k8s-worker-111:~# echo "deb https://apt.kubernetes.io/ kubernetes-xenial main" | sudo tee /etc/apt/sources.list.d/kubernetes.list
deb https://apt.kubernetes.io/ kubernetes-xenial main

root@k8s-worker-111:~# apt update
Hit:1 http://deb.debian.org/debian bullseye InRelease
Hit:2 http://security.debian.org/debian-security bullseye-security InRelease                                                        
Hit:3 http://deb.debian.org/debian bullseye-updates InRelease                                                                       
Hit:4 http://deb.debian.org/debian bullseye-backports InRelease                                               
Hit:6 https://download.docker.com/linux/debian bullseye InRelease                       
Get:5 https://packages.cloud.google.com/apt kubernetes-xenial InRelease [9383 B]
Get:7 https://packages.cloud.google.com/apt kubernetes-xenial/main amd64 Packages [58.4 kB]
Fetched 67.8 kB in 4s (18.8 kB/s)
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
All packages are up to date.
```

By the way, it's allways " https://apt.kubernetes.io/ kubernetes-xenial main" for all Debian based distributions ?
On https://packages.cloud.google.com/apt/dists:
```
kubernetes-jessie
kubernetes-jessie-unstable
kubernetes-lucid
kubernetes-lucid-unstable
kubernetes-precise
kubernetes-precise-unstable
kubernetes-squeeze
kubernetes-squeeze-unstable
kubernetes-stretch
kubernetes-stretch-unstable
kubernetes-trusty
kubernetes-trusty-unstable
kubernetes-wheezy
kubernetes-wheezy-unstable
kubernetes-xenial
kubernetes-xenial-unstable
kubernetes-yakkety
kubernetes-yakkety-unstable
```

**Comment by mrajadurai2413** [comment]
`sudo curl -fsSLo /etc/apt/keyrings/kubernetes-archive-keyring.gpg https://packages.cloud.google.com/apt/doc/apt-key.gpg
`
`echo "deb [signed-by=/etc/apt/keyrings/kubernetes-archive-keyring.gpg] https://apt.kubernetes.io/ kubernetes-xenial main" | sudo tee /etc/apt/sources.list.d/kubernetes.list`

**Comment by narensrini-ds** [comment]
Running the command from @mrajadurai2413 is working and solved the issues for missing public key for Kubernetes
`W: An error occurred during the signature verification. The repository is not updated and the previous index files will be used. GPG error: https://packages.cloud.google.com/apt kubernetes-xenial InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY B53DC80D13EDEF05
W: Failed to fetch https://apt.kubernetes.io/dists/kubernetes-xenial/InRelease  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY B53DC80D13EDEF05
W: Some index files failed to download. They have been ignored, or old ones used instead.
`

**Comment by afolarin** [comment]
`/etc/apt/keyrings/kubernetes-archive-keyring.gpg`
or 
`/usr/share/keyrings/kubernetes-archive-keyring.gpg`

which is correct?

**Comment by jsoref** [comment]
It depends on your definition.

The second one assumes you're the os vendor/distributor, which you probably aren't. 

The first one assumes you're the owner of the machine.

**Comment by jiajie-chen-havas** [comment]
I was able to get this working again on Ubuntu 18.04 using [mrajadurai2413's solution](https://github.com/kubernetes/release/issues/1982#issuecomment-1385479784). 
Just to make sure things were working correctly, I also did `chmod root:root` and `chown 644` on the `.gpg` and `.list` files.

There's some [security concerns](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=851774#31) using `apt-key add` without `--keyfile` set, but `apt-key` is [being deprecated in Debian/Ubuntu anyways](https://manpages.debian.org/bullseye/apt/apt-key.8.en.html).

**Comment by risteloo** [comment]
I have the same version of Ubuntu as @narensrini-ds and the following commands has been working for me.
The correct path for me is /usr/share/keyrings/kubernetes-archive-keyring.gpg 

1) curl -fsSLo /usr/share/keyrings/kubernetes-archive-keyring.gpg 'https://packages.cloud.google.com/apt/doc/apt-key.gpg'
2) echo "deb [signed-by=/usr/share/keyrings/kubernetes-archive-keyring.gpg] https://apt.kubernetes.io/ kubernetes-xenial main" | sudo tee /etc/apt/sources.list.d/kubernetes.list
3) sudo apt update && sudo apt upgrade
4) sudo apt-get update && sudo apt-get upgrade

Thank you so much @narensrini-ds and @afolarin

**Comment by BdN3504** [comment]
To make this work under Debian 5.10.162-1 I adjusted the keyring directories:

1. `sudo curl -fsSLo /etc/apt/trusted.gpg.d/kubernetes-archive-keyring.gpg https://packages.cloud.google.com/apt/doc/apt-key.gpg`
2. ```
echo "deb [signed-by=/etc/apt/trusted.gpg.d/kubernetes-archive-keyring.gpg] https://apt.kubernetes.io/ kubernetes-xenial main" | sudo tee /etc/apt/sources.list.d/kubernetes.list
deb [signed-by=/etc/apt/trusted.gpg.d/kubernetes-archive-keyring.gpg] https://apt.kubernetes.io/ kubernetes-xenial main
```
3. `sudo apt update`

**Comment by jsoref** [comment]
Best practices say: You should *not* be putting files in the `trusted.gpg.d` directory.

**Comment by afolarin** [comment]
I think `trusted.gpg` is not recommended due to  global trust of keys under that keyring by apt opening a security issue. But it is now  recommended to use separate keyrings under `trusted.gpg.d`

See

https://itsfoss.com/apt-key-deprecated/

**Comment by jsoref** [comment]
No. As poorly documented later in that page, the correct way is:
> And then the dearmored key is added to its own file under /usr/share/keyrings directory

(Or a similar directory in /etc.)

The trusted.gpg.d directory has the same problems as the trusted.gpg file – all entries in it are trusted for *all* repositories, unless each repository is directly pinned, at which point there's really no reason to use the directory.

**Comment by afolarin** [comment]
@jsoref looks like I stand corrected - the documentation I linked wasn't the clearest.

https://wiki.debian.org/DebianRepository/UseThirdParty

> "The certificate MUST NOT be placed in `/etc/apt/trusted.gpg.d` or loaded by apt-key add."

**Comment by BdN3504** [comment]
I used the `/etc/apt/trusted.gpg.d` directory because that's what is stated in the [apt-key man page](https://manpages.debian.org/stretch/apt/apt-key.8.en.html):
**add** filename
Add a new key to the list of trusted keys. The key is read from the filename given with the parameter filename or if the filename is - from standard input.

It is critical that keys added manually via apt-key are verified to belong to the owner of the repositories they claim to be for otherwise the [apt-secure(8)](https://manpages.debian.org/stretch/apt/apt-secure.8.en.html) infrastructure is completely undermined.

_Note: Instead of using this command a keyring **should be placed directly in the /etc/apt/trusted.gpg.d/** directory with a descriptive name and either "gpg" or "asc" as file extension._

If this is not the case anymore, I'll put it somewhere appropriately.

**Comment by jsoref** [comment]
If you read the latest version of the man page, instead of expecting the upstream to backport deprecation messages to the version you're using, you'll see that the tool you're using is deprecated:
https://manpages.debian.org/experimental/apt/apt-key.8.en.html

Technically a given version of a platform will generally support (to some limited definition of support) the tools it had when it shipped. That's generally a promise not to *break* functionality in it. Practically, that means you can still use a given program, even if it's a bad idea to do so.

What it doesn't mean and isn't a license to do is to tell others how they should work in the general case.

In the general case, you shouldn't use that program as it's deprecated because it's fundamentally broken.

**Comment by BdN3504** [comment]
Alright, so to correctly set this up for debian I'd suggest the following then:

1. `curl -s https://packages.cloud.google.com/apt/doc/apt-key.gpg | gpg --dearmor | sudo dd status=none of=/usr/share/keyrings/kubernetes-archive-keyring.gpg`
2. `echo "deb [signed-by=/usr/share/keyrings/kubernetes-archive-keyring.gpg] https://apt.kubernetes.io/ kubernetes-xenial main" | sudo tee /etc/apt/sources.list.d/kubernetes.list`
3. sudo apt update

**Comment by jsoref** [comment]
In principle, since you aren't a distribution you should be using an equivalent path in `/etc`, not `/usr`.

**Comment by dims** [comment]
cc @sftim

**Comment by sftim** [comment]
I don't think there are any useful docs changes to make here right now. To really fix things for users, we need to run our own (ie, not Google) package repos, and publicize those.

At the moment I don't have cycles to help with the package repo side. Sorry about that.

**Comment by sftim** [comment]
Are we happy to close this in favor of https://github.com/kubernetes/release/issues/913 ?

**Comment by saschagrunert** [comment]
> Are we happy to close this in favor of ~#913~ [kubernetes/enhancements#1731](https://github.com/kubernetes/enhancements/issues/1731) ?

Yep, let's do this
/close
